nova_vm/src/ecmascript/builtins/global_object.rs

```diff
@@ -112,15 +127,457 @@ impl BuiltinIntrinsic for GlobalObjectUnescape {
+/// language value), strictCaller (a Boolean), and direct (a Boolean) and
+/// returns either a normal completion containing an ECMAScript language value
+/// or a throw completion.
+pub fn perform_eval(
+    agent: &mut Agent,
+    x: Value,
+    direct: bool,
+    strict_caller: bool,
+) -> JsResult<Value> {
+    // 1. Assert: If direct is false, then strictCaller is also false.
+    assert!(direct || !strict_caller);
+
+    // 2. If x is not a String, return x.
+    let Ok(x) = String::try_from(x) else {
+        return Ok(x);
+    };
+
+    // 3. Let evalRealm be the current Realm Record.
+    let eval_realm = agent.current_realm_id();
+
+    // 4. NOTE: In the case of a direct eval, evalRealm is the realm of both the caller of eval and of the eval function itself.
+    // 5. Perform ? HostEnsureCanCompileStrings(evalRealm, « », x, direct).
+    agent
+        .host_hooks
+        .host_ensure_can_compile_strings(&mut agent[eval_realm])?;
+
+    // 6. Let inFunction be false.
+    let mut in_function = false;
+    // 7. Let inMethod be false.
+    let mut in_method = false;
+    // 8. Let inDerivedConstructor be false.
+    let mut in_derived_constructor = false;
+    // 9. Let inClassFieldInitializer be false.
+    let mut in_class_field_initializer = false;
+
+    // 10. If direct is true, then
+    if direct {
+        // a. Let thisEnvRec be GetThisEnvironment().
+        let this_env_rec = get_this_environment(agent);
+        // b. If thisEnvRec is a Function Environment Record, then
+        if let EnvironmentIndex::Function(this_env_rec) = this_env_rec {
+            // i. Let F be thisEnvRec.[[FunctionObject]].
+            let f = agent[this_env_rec].function_object;
+            // ii. Set inFunction to true.
+            in_function = true;
+            // iii. Set inMethod to thisEnvRec.HasSuperBinding().
+            in_method = this_env_rec.has_super_binding(agent);
+            // iv. If F.[[ConstructorKind]] is derived, set inDerivedConstructor to true.
+            in_derived_constructor = match f {
+                Function::ECMAScriptFunction(idx) => agent[idx]
+                    .ecmascript_function
+                    .constructor_status
+                    .is_derived_class(),
+                _ => todo!(),
+            };
+
+            // TODO:
+            // v. Let classFieldInitializerName be F.[[ClassFieldInitializerName]].
+            // vi. If classFieldInitializerName is not empty, set inClassFieldInitializer to true.
+        }
+    }
+
+    // 11. Perform the following substeps in an implementation-defined order, possibly interleaving parsing and error detection:
+    // a. Let script be ParseText(x, Script).
+    let allocator = Allocator::default();
+    let source_text = x.as_str(agent).to_owned();
+    let parser = Parser::new(&allocator, &source_text, SourceType::default());
+    let ParserReturn {
+        errors,
+        program: script,
+        ..
+    } = parser.parse();
```

Comment:
**todo:** `ParseText` still needs to be implemented.

Reply:
I think `ParseText` is just this `parser.parse()`, ie. turn the source string into an AST.

Reply:
Yep pretty much, but it also handles any exceptions which may have been thrown and converts them to `SyntaxError`s.